llama-index-integrations/llms/llama-index-llms-nvidia/llama_index/llms/nvidia/base.py

```diff
@@ -0,0 +1,379 @@
+    astream_chat_to_completion_decorator,
+    chat_to_completion_decorator,
+    get_from_param_or_env,
+    stream_chat_to_completion_decorator,
+)
+from llama_index.core.llms.llm import LLM
+from llama_index.core.types import BaseOutputParser, PydanticProgramMode
+
+from llama_index.llms.nvidia.utils import (
+    catalog_modelname_to_contextsize,
+    API_CATALOG_MODELS,
+)
+
+from llama_index.llms.openai.utils import (
+    from_openai_message,
+    to_openai_message_dicts,
+)
+
+from openai import OpenAI as SyncOpenAI
+from openai import AsyncOpenAI
+
+DEFAULT_CATALOG_MODEL = "meta/llama3-8b-instruct"
+BASE_CATALOG_URL = "https://integrate.api.nvidia.com/v1/"
+DEFAULT_CATALOG_MAX_TOKENS = 512
+
+
+class Model(BaseModel):
+    id: str
+
+
+class NVIDIA(LLM):
+    """NVIDIA's API Catalog Connector."""
+
+    model: str = Field(
+        default=DEFAULT_CATALOG_MODEL,
+        description="The NVIDIA API Catalog model to use.",
+    )
+    temperature: float = Field(
+        default=DEFAULT_TEMPERATURE,
+        description="The temperature to use for sampling.",
+        gte=0.0,
+        lte=1.0,
+    )
+    max_tokens: int = Field(
+        default=DEFAULT_CATALOG_MAX_TOKENS,
+        description="The maximum number of tokens to generate.",
+        gte=0,
+    )
+
+    timeout: float = Field(
+        default=120, description="The timeout for the API request in seconds.", gte=0
+    )
+
+    max_retries: int = Field(
+        default=5,
+        description="The maximum number of retries for the API request.",
+        gte=0,
+    )
+
+    _client: Any = PrivateAttr()
+    _aclient: Any = PrivateAttr()
+    _mode: str = PrivateAttr("nvidia")
+
+    def __init__(
+        self,
+        model: str = DEFAULT_CATALOG_MODEL,
+        temperature: float = DEFAULT_TEMPERATURE,
+        max_tokens: int = DEFAULT_CATALOG_MAX_TOKENS,
+        timeout: float = 120,
+        max_retries: int = 5,
+        nvidia_api_key: Optional[str] = None,
+        api_key: Optional[str] = None,
+        callback_manager: Optional[CallbackManager] = None,
+        system_prompt: Optional[str] = None,
+        messages_to_prompt: Optional[Callable[[Sequence[ChatMessage]], str]] = None,
+        complettion_to_prompt: Optional[Callable[[str], str]] = None,
+        pydantic_program_mode: PydanticProgramMode = PydanticProgramMode.DEFAULT,
+        output_parser: Optional[BaseOutputParser] = None,
+    ) -> None:
+        callback_manager = callback_manager or CallbackManager([])
+
+        api_key = get_from_param_or_env(
+            "api_key",
+            nvidia_api_key or api_key,
+            "NVIDIA_API_KEY",
+            "NO_API_KEY_PROVIDED",
+        )
+
+        self._client = SyncOpenAI(
+            api_key=api_key,
+            base_url=BASE_CATALOG_URL,
+            timeout=timeout,
+            max_retries=max_retries,
+        )
+        self._client._custom_headers = {"User-Agent": "llama-index-llms-nvidia"}
+        self._aclient = AsyncOpenAI(
```

Comment:
Do you think it's worth subclassing `OpenAILike` instead? Seems like all the logic is basically the same except for the `init` ?

Reply:
TIL `OpenAILike` :bow:

i just did a prototype using `OpenAILike` instead of having Async/OpenAI clients. i ran into an issue with top_logprobs ('NoneType' object is not iterable) in from_openai_token_logprobs. this suggests to me that the input may be openai-like but the output isn't openai-like enough for this to work yet.

Reply:
Hmm its weird that you got the error -- the outputs and inputs should be consistent since its all using the same openai client 🤔

Reply:
I imagine other users will also need that patch haha

Reply:
@logan-markewich i've pushed an alternative implementation using OpenAILike. i changed your fix for the None logprobs to be part of the OpenAI util and added tests.

when someone makes a change to a package they're obligated to run that package's tests. with l-i-llm-nvidia depending on l-i-llm-openai, how can we ensure changes to l-i-llm-openai don't break l-i-llm-nvidia?